cmake/build_helpers.cmake

```diff
@@ -11,10 +11,13 @@ function(ginkgo_default_includes name)
         )
     if(GINKGO_HAVE_HWLOC)
       target_include_directories("${name}"
-        PUBLIC 
+        PUBLIC
         $<BUILD_INTERFACE:${HWLOC_INCLUDE_DIRS}>
         )
     endif()
+    if(GINKGO_BUILD_MPI)
+        target_include_directories("${name}" PUBLIC ${MPI_CXX_INCLUDE_DIRS})
```

Comment:
Does the target include `mpi.h` itself? Then maybe
```suggestion
        target_link_libraries("${name}" PUBLIC MPI::MPI_CXX)
```
Otherwise this should be propagated from the `ginkgo` target

Reply:
We don't call any MPI functions within the modules, so I didn't want to use `target_link_libraries`. But we now have to include `mpi.h` basically everywhere, because it is included from `polymorphic_object.hpp`. So we need to explicitly either add the include paths or link against the mpi target for all of our modules.
I would still prefer to only include the header, but I'm fine with linking against mpi.

Reply:
I still don't see why this is necessary: Ginkgo links against `MPI::MPI_CXX` publically, so the dependency should be propagated to any target that includes `polymorphic_object.hpp`. Is something broken in our CMake setup?

Reply:
Our core module is linked against MPI::MPI_CXX, but nothing else (except tests). So our modules are not linked against MPI, and honestly I would prefer if we could keep it that way. But due to the changes in `polymorphic_object.hpp`, now every module needs to have the MPI include path available.

Reply:
I could move the new classes into separate files, and that way we might be able to circumvent all of this.

Reply:
Anything that links against `ginkgo` should also inherit the dependency on MPI_CXX, does that not happen?

Reply:
The modules are only linked against `ginkgo_device`. For the same reason we can't call core functions within the modules.

Reply:
In that case, this looks fine. Ideally, every header file should belong only to a single target, and linking against that target should pull in all dependencies. ATM we can't guarantee that yet.

Reply:
I've moved the classes into separate files, so now we don't need to include MPI in the modules anymore.

Reply:
Nice, thanks!